infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/CorfuReplicationDiscoveryService.java

```diff
@@ -419,11 +418,21 @@ public void processLockRelease() {
      * @param newTopology new discovered topology
      */
     public void processClusterRoleChange(TopologyDescriptor newTopology) {
+        // stop ongoing replication, stopLogReplication() checks leadership and active
         stopLogReplication();
+
         //TODO pankti: read the configuration again and refresh the LogReplicationConfig object
-        replicationManager.setTopology(newTopology);
-        boolean activeCluster = localNodeDescriptor.getRoleType() == ClusterRole.ACTIVE;
-        updateTopologyConfigId(activeCluster);
+
+        // update topology, cluster, and node configs
+        updateLocalTopology(newTopology);
+
+        // update config id in metadata manager
+        interClusterReplicationService.getLogReplicationServer().getSinkManager()
+                .updateTopologyConfigId(topologyDescriptor.getTopologyConfigId());
```

Comment:
I think updateTopologyConfigId is not the correct naming in the SinkManager, right? Cause that method has added functionality and is really doing much more than that... maybe processClusterRoleChange or resetOnClusterRoleChange?

Reply:
Also, similar to stop log replication isn't this dependent of the fact that you are currently a standby?... In such case, it might be a good idea to put this statement in the stopLogReplication method in the case STANDBY, right?

Reply:
The function of updateTopologyCongId will update the metadata and sinkManager related information. It should be called on all the type of nodes.

Reply:
But only the metadata, right? however, not the snapshotWriter/logEntryWriter. 
Those actually have logic that include calls to the database which are not required if you're not the lead standby, as a non-lead is not really doing anything. There is access to the logReplicationMedataManager at the DiscoveryService level, we can remove the updateTopologyConfigId on the logReplicationMetadata from the sink so it applies to all nodes, and the reset of the writers only if you're the leader, or is there something else that is generic to leaders and non-leaders?

Reply:
1. For all the nodes, it needs to update the metadata if config changes. As the metadata is protected by the transaction, at each site, only one node will succeed.
2. The related writer state should be updated too. As the configuration changes, a new full snapshot full sync will be needed. This is to reset the writer state. It is safer to reset writer state at all nodes as the leadership can change at anytime. In my local PR, I am trying to make the writer stateless and only using the persistent metadata.

Reply:
Assume the case that, there is a leadership change from node1 to node2. The leadership acquire event is on the queue for node2's discovery service to process, so node2 think it is not leader. Node1 has got the notification of leadership loss and has processed it and node1 think it is not leader either. There are some intermittent time, there is no node that consider it is the  leader and we don't want to lose the update of the metadata with the new config.

Reply:
Ok ok, yes in that case even though it is redundant it will protect us in these cases.